environment: Replace LD with <LANG>LD

dcbaker opened the pull request:
The rust code is ugly, because rust is annoying. It doesn't invoke a
linker directly (unless that linker is link.exe or lld-link.exe),
instead it invokes the C compiler (gcc or clang usually) to do it's
linking. Meson doesn't have good abstractions for this, though we
probably should because some of the D compilers do the same thing.
Either that or we should just call the c compiler directly, like vala
does.

This changes the public interface for meson, which we don't do unless we
absolutely have to. In this case I think we need to do it. A fair number
of projects have already been using 'ld' in their cross/native files to
get the ld binary and call it directly in custom_targets or generators,
and we broke that. While we could hit this problem again names like
`cld` and `cppld` are far less likely to cause collisions than `ld`.
Additionally this gives a way to set the linker on a per-compiler basis,
which is probably in itself very useful.

Fixes #6442

ignatenkobrain commented on mesonbuild/environment.py:
```diff
@@ -1378,32 +1379,45 @@ def detect_rust_compiler(self, for_machine):
+                                                **extra_args)
+                elif 'link' in override[0]:
+                    linker = self._guess_win_linker(
+                        override, RustomCompiler, for_machine, use_linker_prefix=False)
```
__[Flake8]__ undefined name 'RustomCompiler' ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6457#issue-5057270))

<details>

| Rule |
| - |
| F821 |

</details>
<sub>posted by Sider</sub>

ignatenkobrain commented on mesonbuild/environment.py:
```diff
@@ -1378,32 +1379,45 @@ def detect_rust_compiler(self, for_machine):
+                    else:
+                        linker = type(cc.linker)(compiler, for_machine, cc.linker.id, cc.LINKER_PREFIX,
+                                                always_args=always_args, version=cc.linker.version,
+                                                **extra_args)
```
__[Flake8]__ continuation line under-indented for visual indent ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6457#issue-5057269))

<details>

| Rule |
| - |
| E128 |

</details>
<sub>posted by Sider</sub>

ignatenkobrain commented on mesonbuild/environment.py:
```diff
@@ -1378,32 +1379,45 @@ def detect_rust_compiler(self, for_machine):
+                    # so we can initialize a new copy for the Rust Compiler
+                    if is_link_exe:
+                        linker = type(cc.linker)(for_machine, always_args, exelist=cc.linker.exelist,
+                                                version=cc.linker.version, **extra_args)
```
__[Flake8]__ continuation line under-indented for visual indent ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6457#issue-5057267))

<details>

| Rule |
| - |
| E128 |

</details>
<sub>posted by Sider</sub>

ignatenkobrain commented on mesonbuild/environment.py:
```diff
@@ -1378,32 +1379,45 @@ def detect_rust_compiler(self, for_machine):
+                                                version=cc.linker.version, **extra_args)
+                    else:
+                        linker = type(cc.linker)(compiler, for_machine, cc.linker.id, cc.LINKER_PREFIX,
+                                                always_args=always_args, version=cc.linker.version,
```
__[Flake8]__ continuation line under-indented for visual indent ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6457#issue-5057268))

<details>

| Rule |
| - |
| E128 |

</details>
<sub>posted by Sider</sub>

ignatenkobrain commented on mesonbuild/environment.py:
```diff
@@ -1378,32 +1379,45 @@ def detect_rust_compiler(self, for_machine):
+                        extra_args['direct'] = True
+                        extra_args['machine'] = cc.linker.machine
+                    elif not ((info.is_darwin() and isinstance(cc, AppleClangCCompiler)) or
+                            isinstance(cc, GnuCCompiler)):
```
__[Flake8]__ continuation line under-indented for visual indent ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6457#issue-5057266))

<details>

| Rule |
| - |
| E128 |

</details>
<sub>posted by Sider</sub>

jmesmon commented on docs/markdown/howtox.md:
````diff
@@ -28,24 +28,36 @@ native-files and the latter via the cross file only.
+```ini
+[binaries]
+c = 'clang'
+cld = 'lld'
````
Is there discussion somewhere about the choice to use a `binaries` entry for something that isn't the name of a program?

To be clear: `lld` here happens to be a program name and a linker identifier. In many cases, though, this entry would not be a program name at all (for example, to pick `bfd` when using gcc).

It seems reasonable things that aren't binaries shouldn't be in the `binaries` section. Something like https://github.com/mesonbuild/meson/issues/6442#issuecomment-573302028 seems like a reasonable approach if we want to specify a program.

dcbaker replied:
This is an oddity of how meson is implemented. For some compiler/linker combinations we drive the linker through the compiler, others we don't. We invoke gnu ld and apple ld through gcc or clang, but we invoke link.exe and lld-link.exe directly. Having two different ways to do the same thing (pick which linker) seems really annoying and prone to "Why does this not work?" kind of issues.

dcbaker replied:
And also, "why do I put the compiler in binaries, but the linker in a different section?" Is a total valid argument. If we were doing this from scratch with what we know now, we would probably split the "binaries" section into four separate sections, a "compilers", a "linkers", a "binaries", and a "programs", where "binaries" is specifically for meson internal binaries pkg-config, cmake, llvm-config, etc, and "programs" is for things that `find_program` looks up.

Arfrever replied:
Regardless of section name, as suggested [here](https://github.com/mesonbuild/meson/issues/6442#issuecomment-573231412), could better names of new variables be used?
C and C++ compiler are specified in cross file by `c` and `cpp` variables, and their arguments by `c_args` and `cpp_args`, **not** `cargs` and `cppargs`, so names `cld` and `cppld` would be inconsistent. Please see [this](https://mesonbuild.com/Reference-tables.html#language-arguments-parameter-names).

Since `bfd` or `gold` are linker **identifiers**, and not exact filenames, I suggest to include `id` / `ID` in names of new variables.

Arguments passed to linkers are already specified in `c_link_args`, `cpp_link_args` etc., so something similar would make sense.

For `pkgconfig`, environment variable is `PKG_CONFIG`, so using underscore is not unprecedented.

Summary of new variables (in cross file and environment) suggested by me:
* `c_linker_id`: `C_LD_ID`
* `cpp_linker_id`: `CXX_LD_ID`
* `cs_linker_id`: `CS_LD_ID` (If it makes sense... A pair of variables for this language was omitted in original patch.)
* `d_linker_id`: `D_LD_ID`
* `fortran_linker_id`: `F_LD_ID`
* `objc_linker_id`: `OBJC_LD_ID`
* `objcpp_linker_id`: `OBJCXX_LD_ID`
* `rust_linker_id`: `RUST_LD_ID`

jpakkane commented:
Some quick comments:

There should be a test that does something like this under linuxlike:

```meson
ld = find_program('ld')
assert(run_command(ld, '--version').returncode() == 0, 'Linker lookup not working.')
```

I agree that the option names should be `c_ld` instead of `cld`.

Could we make it do the thing I mentioned in the bug where the ld entries point to binaries rather than an identifier string? Then the code would do:

- run something like `ldbinary --version`
- grep the linker type, store it in a variable in the linker
- when linking through the compiler, pass the linker id to the compiler and have it add the flags needed to use said linker (i.e. when using Gold, add `-fuse-ld=gold`)

The detection logic would probably be slightly wrong at first, but as long as the basic setups work we can improve the detection logic as people keep filing bugs.

dcbaker commented:
> I agree that the option names should be c_ld instead of cld.

I can do that

> Could we make it do the thing I mentioned in the bug where the ld entries point to binaries rather than an identifier string? Then the code would do:
>
>    run something like ldbinary --version
      grep the linker type, store it in a variable in the linker
      when linking through the compiler, pass the linker id to the compiler and have it add the flags needed to use said linker (i.e. when using Gold, add -fuse-ld=gold)


Maybe I'm misunderstanding, but what it sounds like your saying is:

```ini
[binaries]
c_ld = '/bin/ld.gold'
```

Then we would call `ld.gold --version`, but would know that anything that ends with `ld.gold` really means `-fuse-ld=gold`.

How in that case do we handle a case like:

```ini
[binaries]
c = '/bin/gcc'
c_ld = '/opt/bin/ld.gold`
```
AFAICT neither gcc nor clang have a method for invoking a linker that isn't the one that was set at configure/compile time. They are simply hard coded. By letting people put in paths we're basically lying to them about what we can do.
```ini

dcbaker commented:
I've updated with  the 'cld' -> 'c_ld' changes requested by @jpakkane, I've also changed the environment variables the same way `CCLD` -> `CC_LD`, though I'm considering just dropping them altogether since meson by design can't really do what autotools does. I've also fixed the sider issues.

I'll write the linuxlike test next.

minerva commented:
> > I agree that the option names should be c_ld instead of cld.
> 
> I can do that
> 
> > Could we make it do the thing I mentioned in the bug where the ld entries point to binaries rather than an identifier string? Then the code would do:
> > run something like ldbinary --version
> > grep the linker type, store it in a variable in the linker
> > when linking through the compiler, pass the linker id to the compiler and have it add the flags needed to use said linker (i.e. when using Gold, add -fuse-ld=gold)
> 
> Maybe I'm misunderstanding, but what it sounds like your saying is:
> 
> ```ini
> [binaries]
> c_ld = '/bin/ld.gold'
> ```
> 
> Then we would call `ld.gold --version`, but would know that anything that ends with `ld.gold` really means `-fuse-ld=gold`.
> 
> How in that case do we handle a case like:
> 
> ```ini
> [binaries]
> c = '/bin/gcc'
> c_ld = '/opt/bin/ld.gold`
> ```
> 
> AFAICT neither gcc nor clang have a method for invoking a linker that isn't the one that was set at configure/compile time. They are simply hard coded. By letting people put in paths we're lying to them about what we can do.

I'm not sure about gcc, but clang accepts a full path leading to the linker binary used by the compiler driver on recent versions, not just an identifier.

dcbaker commented:
> I'm not sure about gcc, but clang accepts a full path leading to the linker binary used by the compiler driver on recent versions, not just an identifier.
>
> Edit: https://reviews.llvm.org/D17952

This must be in clang 9 or 10, but clang 8 doesn't support that. It's good to know that clang is different in this regard.

jpakkane commented:
> Then we would call ld.gold --version, but would know that anything that ends with ld.gold really means -fuse-ld=gold.

You can tell the difference between `ld --version` and `ld.gold --version` so one could detect that. But upon thinking more about it maybe that can get too complicated, especially since `/usr/bin/ld` can be either bfd or gold. So maybe doing this is the smart thing to do.

If it turns out that this causes problems, we can change the semantics in the future so that if you give an absolute path, we do the detection and if you give a plain string (i.e. `gold`), then we do the thing we do currently.

jmesmon commented:
> If it turns out that this causes problems, we can change the semantics in the future so that if you give an absolute path, we do the detection and if you give a plain string (i.e. gold), then we do the thing we do currently.

I'd recommend against requiring absolute paths here. Plain executable names (like `ld = "arm-none-eabi-ld"`) are what I use in my cross files and it seems likely others do to.

jpakkane commented:
The test failures seem to be caused by out of date dictionaries:

```
>       envvar = mesonbuild.envconfig.BinaryTable.evarMap['{}ld'.format(lang)]
E       KeyError: 'cppld'
```

jpakkane commented:
The key failure can be fixed with this:

```patch
diff --git a/run_unittests.py b/run_unittests.py
index c87e280d..880579d4 100755
--- a/run_unittests.py
+++ b/run_unittests.py
@@ -5880,7 +5880,7 @@ c = ['{0}']
             raise unittest.SkipTest('Solaris currently cannot override the linker.')
         if not shutil.which(check):
             raise unittest.SkipTest('Could not find {}.'.format(check))
-        envvar = mesonbuild.envconfig.BinaryTable.evarMap['{}ld'.format(lang)]
+        envvar = mesonbuild.envconfig.BinaryTable.evarMap['{}_ld'.format(lang)]
         with mock.patch.dict(os.environ, {envvar: name}):
             env = get_fake_env()
             comp = getattr(env, 'detect_{}_compiler'.format(lang))(MachineChoice.HOST)
```

However the actual linker detection fails. The environment variable is not used for the lookup and the default is used:

```
$ CC_LD=gold ./meson.py test\ cases/common/1\ trivial build

The Meson build system
...
C compiler for the host machine: ccache cc (gcc 9.2.1 "cc (Ubuntu 9.2.1-9ubuntu2) 9.2.1 20191008")
C linker for the host machine: cc GNU ld.bfd 2.33
```

dcbaker commented:
> However the actual linker detection fails. The environment variable is not used for the lookup and the default is used

Because the exact same staleness was present in the actual code :/

jpakkane commented on test cases/linuxlike/15 ld binary/meson.build:
```diff
@@ -0,0 +1,4 @@
+project('ld binary')
+
+ld = find_program('ld')
+assert(run_command(ld, '--version').returncode() == 0)
```
I just realized that the cross files in `cross` subdirectory should have ld entries too. Maybe add them to armhf and mingw32 and leave it out from mingw64 so that both cases get tested?

dcbaker replied:
done. I'm waiting for my windows machine to finish running updates, then I'll try to sort those out.
